File: packages/battle_engine/lib/src/ability_effect.dart
import 'field_effect.dart';
import 'targeted_status.dart';

/// The accumulated effect of resolving an [Ability]'s mutations, before it
/// is applied to a [BattleState].
///
/// [hitCount] and [critChanceBonus] are not consumed by anything yet — there
/// is no damage/crit system in the engine. They are carried as data for
/// whichever system implements that later, rather than being computed here.
class AbilityEffect {
  final List<TargetedStatus> statusesToApply;
  final FieldEffect? fieldEffect;
  final int hitCount;
  final double critChanceBonus;

  const AbilityEffect({
    this.statusesToApply = const [],
    this.fieldEffect,
    this.hitCount = 1,
    this.critChanceBonus = 0,
  });

  AbilityEffect copyWith({
    List<TargetedStatus>? statusesToApply,
    FieldEffect? fieldEffect,
    int? hitCount,
    double? critChanceBonus,
  }) {
    return AbilityEffect(
      statusesToApply: statusesToApply ?? this.statusesToApply,
      fieldEffect: fieldEffect ?? this.fieldEffect,
      hitCount: hitCount ?? this.hitCount,
      critChanceBonus: critChanceBonus ?? this.critChanceBonus,
    );
  }
}

File: backend/src/battle-rules/ability-effect.ts
import type { ActiveStatus, FieldEffect } from "./types.js";

/** Mirrors StatusTarget in targeted_status.dart — who a mutation-applied
 * status lands on. Most mutations (e.g. Combustão) debuff the opponent —
 * the default in Dart; a defensive one (e.g. Guarda) needs to protect
 * whoever's using it instead. */
export type StatusTarget = "actor" | "opponent";

/** Mirrors TargetedStatus in battle_engine (Dart). */
export interface TargetedStatus {
  readonly status: ActiveStatus;
  readonly target: StatusTarget;
}

/** Mirrors AbilityEffect in battle_engine (Dart) — the accumulated effect
 * of resolving an ability's mutations, before it's applied to a
 * BattleState. `hitCount`/`critChanceBonus` from the Dart source aren't
 * mirrored: nothing on either side consumes them yet (see the doc comment
 * on AbilityEffect in ability_effect.dart), so carrying them here would be
 * dead weight — see DECISION-025. */
export interface AbilityEffect {
  readonly statusesToApply: readonly TargetedStatus[];
  readonly fieldEffect: FieldEffect | null;
}

export const emptyAbilityEffect: AbilityEffect = {
  statusesToApply: [],
  fieldEffect: null,
};
